docs: add documents for the data structure of Cell/Script/Transaction/Block (#309)
nervosnetwork/ckb: docs/data-structures.md

## Transaction

### Example

```json
{
    "deps": [],
    "hash": "0x4707810253259258e3091934dc5b543403c10cc899859e077fe26067f8d52dc0",
    "inputs": [
      {
        "previous_output": {
          "hash": "0x0000000000000000000000000000000000000000000000000000000000000000",
          "index": 4294967295
        },
        "unlock": {
          "args": [],
          "binary": "0x0d00000000000000",
          "reference": null,
          "signed_args": [],
          "version": 0
        }
      }
    ],
    "outputs": [
      {
        "capacity": 5000000,
        "data": "0x",
        "lock": "0xa58a960b28d6e283546e38740e80142da94f88e88d5114d8dc91312b8da4765a",
        "type": null
      }
    ],
    "version": 0
}
```

### Description

| Name                     | Type         | Description                                                  |
| ------------------------ | ------------ | ------------------------------------------------------------ |
| `version`                | uint32       | **The version of the transaction.** It‘s used to distinguish transactions when there's a fork happened to the blockchain system. |

> | `deps`                   | [`cell`]     | **An array of cells that are dependencies of this transaction.** Only live cells can be listed here. The cells listed are read-only. |

Reviewer: Type is `[out_point]`

> | `inputs.previous_output` | [`outpoint`] | **An array of cell outpoints that point to the cells used as inputs.** Input cells are in fact the output of previous transactions, hence they are noted as `previous_output` here. These cells are referred through  `outpoint`, which contains the transaction `hash` of the previous transaction, as well as this cell's `index` in its transaction's output list. |

Reviewer: `inputs` is an array of  `{previous_output, unlock}`

# Data Structures of Cell/Script/Transaction/Block

This documents list all the basic structures one may need to know in order to develop on CKB, or mine CKB coins. 

* [Cell](#Cell)
* [Script](#Script)
* [Transaction](#Transaction)
* [Block](#Block)

## Cell

### Example

```json
{
    "capacity": 5000000,
    "data": "0x",
    "lock": "0xa58a960b28d6e283546e38740e80142da94f88e88d5114d8dc91312b8da4765a",
    "type": null
}
```

## Description

| Name       | Type       | Description                                                  |
| :--------- | :--------- | :----------------------------------------------------------- |

> | `capacity` | uint64     | **The size of the cell.** When a new cell is generated (via transaction), one of the verification rule is `capacity = len(capacity)+len(data)+len(type)+len(lock)`. This value also represents the balance of CKB coin, just like the `balance` field in the Bitcoin's UTXO. (E.g. Alice owns 100 CKB coins means she can unlock a group of cells that has 100 amount of `capacity` in total.) |

Reviewer 2: The rule is `capacity ≥ len(capacity)+len(data)+len(type)+len(lock)`

> | `data`     | Bytes      | **Arbitrary data.** This part is for storing states or scripts.  In order to make this cell valid on-chain, the data filled in this field should comply with the logics and rules defined by `type` or `lock`. |

Reviewer 3: ```suggestion
| `data`     | Bytes      | **Arbitrary data.** This part is for storing states or scripts.  In order to make this cell valid on-chain, the data filled in this field should comply with the logic and rules defined by `type`. |
```

The data field is only constrained by the `type`, The `lock` defines the ownership of the cell.

Author: It also could be constrained by `lock`, as `type` and `lock` both has access to all the information of the transaction.

Reviewer 3: I see.

> | `type`     | `Script`   | **A Script that defines the type of the cell.** In a transaction, if an input cell and an output cell has the same `type` field, then the `data` part of these two cells is limited by the `type` script upon the transaction verification. (I.e. `type` is a script that limits how the `data` field of the new cells can be changed from the old cells.) `type` is required to has a data structure of `script`. This field can be empty in a Cell. |

Reviewer 2: The description is not accurate.

@xxuejie Could you help to give a description of `type`?

Reviewer 3: `type` is an option field, it should be explicit to explain.

Reviewer 4: `A script used to enforce certain cell properties. Typical use cases including user defined token support, cell data format validation, etc.`

Author: done

| `lock`     | H256(hash) | **The hash of a Script that defines the ownership of the cell**, just like the `lock` field in the Bitcoin's UTXO. Whoever can provide an unlock script that has the same hash of a cell's `lock` hash can use this cell as input in an transaction (i.e. has the ownership of this cell). This is similar to the P2SH scheme in Bitcoin. |

> More information about Cell can be found in the [whitepaper](https://github.com/nervosnetwork/rfcs/blob/afe50463bb620393b179bd8f08c263b78e366ab3/rfcs/0002-ckb/0002-ckb.md#42-cell).

Reviewer 3: ```suggestion
More information about Cell can be found in the [whitepaper](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0002-ckb/0002-ckb.md#42-cell).
```

## Script

### Example

```json
{
  "version": 0,
  "reference": "0x12b464bcab8f55822501cdb91ea35ea707d72ec970363972388a0c49b94d377c",
  "signed_args": [
    "024a501efd328e062c8675f2365970728c859c592beeefd6be8ead3d901330bc01"
  ],
  "args": [
    "3044022038f282cffdd26e2a050d7779ddc29be81a7e2f8a73706d2b7a6fde8a78e950ee0220538657b4c01be3e77827a82e92d33a923e864c55b88fd18cd5e5b25597432e9b",
    "1"
  ]
}
```

### Description

| Name          | Type    | Description                                                  |
| :------------ | :------ | :----------------------------------------------------------- |
| `version`     | uint8   | **The version of the script.** It‘s used to distinguish transactions when there's a fork happened to the blockchain system. |
| `binary`      | Bytes   | **ELF formatted binary that contains an RISC-V based script.** This part of data is loaded into an CKB-VM instance when they are specified upon the transaction verification. |

> | `reference`   | Bytes   | **The hash of the script that is referred by this script.** It is possible to refer the script in another cell on-chain as the binary code in this script, instead of entering the binary directly into the script. **Notice:** This is part only works when the `binary` field is empty. |

Reviewer 2: How the cell is referred to here?

@xxuejie can provide help here.

Reviewer 3: https://github.com/nervosnetwork/ckb/blob/develop/core/src/script.rs#L13

| `args`        | [Bytes] | **An array of arguments for the script.** The arguments here are imported into the CKB-VM instance as input arguments for the scripts. This part is NOT used when calculating the hash of the script. |

> | `signed_args` | [Bytes] | **An array of signed arguments for the script.** The arguments with signatures. The `signed_args` and the `args` will be connected into a single vector, and imported into the CKB-VM instance as input arguments. |

Reviewer 3: Most of the time, we put `pubkey` in the `signed_args` and put the signature in `args`. The description of `The arguments with signatures.` has made me confused before, I'm not sure it's clear enough.

Reviewer 2: Yes, *The arguments with signatures* is confusing, they are indeed *arguments included when computing signatures*.

Author: I explained this in [another doc](https://github.com/Mine77/ckb-demo-ruby-sdk/blob/docs/update-docs/docs/how-to-write-contracts.md#script-model), maybe I should explain less here and just refer this doc directly.

> | `signed_args` | [Bytes] | **An array of arguments that belongs to the script for improving code reuse rate**. Please refer [this document](https://github.com/Mine77/ckb-demo-ruby-sdk/blob/docs/update-docs/docs/how-to-write-contracts.md#script-model) for more explanation about this field. |

Reviewer 4: This sounds weird to me(`improving code reuse` part), but I don't have a good suggestion, any thoughts @doitian ?

Reviewer 2: If it is weird, I would suggest just delete it until we have a better understanding and a brief description available.

Author: I think it's ok to leave it this way.

At least it's just weird to you, not causing confusions. 

It indeed is increasing the code reuse rate.

> More information about Script can be [here](https://github.com/nervosnetwork/ckb-demo-ruby-sdk/blob/27669cd6b4f56f8977b725fc0e6582b288ab2b82/docs/how-to-write-contracts.md#script-model).

Reviewer 3: ```suggestion
More information about Script can be [here](https://github.com/nervosnetwork/ckb-demo-ruby-sdk/blob/develop/docs/how-to-write-contracts.md#script-model).
```

> More information about Script can be [here](https://github.com/nervosnetwork/ckb-demo-ruby-sdk/blob/master/docs/how-to-write-contracts.md#script-model). Also you can find how the `Script` structure is implemented by reading [these codes](https://github.com/nervosnetwork/ckb/blob/master/core/src/script.rs#L13).

Reviewer 4: What about using links like this: https://github.com/nervosnetwork/ckb/blob/17b48a0777753a91691146b3817758c02db6f415/core/src/script.rs#L13 to ensure future changes won't put this link to a line that is not relevant?

## Transaction

### Example

```json
{
    "deps": [],
    "hash": "0x4707810253259258e3091934dc5b543403c10cc899859e077fe26067f8d52dc0",
    "inputs": [
      {
        "previous_output": {
          "hash": "0x0000000000000000000000000000000000000000000000000000000000000000",
          "index": 4294967295
        },
        "unlock": {
          "args": [],
          "binary": "0x0d00000000000000",
          "reference": null,
          "signed_args": [],
          "version": 0
        }
      }
    ],
    "outputs": [
      {
        "capacity": 5000000,
        "data": "0x",
        "lock": "0xa58a960b28d6e283546e38740e80142da94f88e88d5114d8dc91312b8da4765a",
        "type": null
      }
    ],
    "version": 0
}
```

### Description

#### Transaction

| Name              | Type                             | Description                                                  |
| ----------------- | -------------------------------- | ------------------------------------------------------------ |

> | `version`         | uint32                           | **The version of the transaction.** It‘s used to distinguish transactions when there's a fork happened to the blockchain system. |

Reviewer 3: ```suggestion
| `version`         | uint32                           | **The version of the transaction.** It‘s used to distinguish transactions when there's a fork happened to the blockchain. |
```

Author: "blockchain" is just a data structure. 

"blockchain system" is the system that utilize a blockchain data structure.

| `deps`            | [`outpoint`]                     | **An array of `outpoint` that point to the cells that are dependencies of this transaction.** Only live cells can be listed here. The cells listed are read-only. |
| `inputs`          | [{`previsou_output` , `unlock`}] | **An array of {`previsou_output` , `unlock`}.**              |

> | `previous_output` | [`outpoint`]                     | **An array of cell outpoints that point to the cells used as inputs.** Input cells are in fact the output of previous transactions, hence they are noted as `previous_output` here. These cells are referred through  `outpoint`, which contains the transaction `hash` of the previous transaction, as well as this cell's `index` in its transaction's output list. |

Reviewer 2: `previous_output` is a `outpoint`, not an array of outpoints

> | `unlock`          | `script`                         | **A script for unlocking the corresponding input cell** (i.e. `previous_output`). See [here](https://github.com/nervosnetwork/ckb-demo-ruby-sdk/blob/develop/docs/how-to-write-contracts.md) for how to program this part. |

Reviewer 3: The `previous_output` and `unlock` should not be included, they are part of `inputs`, not the transaction.

Author: The fact that `inputs` is part of transaction and `previsou_output` and `unlock` is part of `input` makes them a part of `inputs` as well.

A keyboard is a part of a laptop, which does not make any of the keys is not part of it.

So I don't think this part could cause too much confusion.

More importantly, this docs serve as a dictionary for looking up parameters, its more convenient  to put related things near each other.

> | `unlock`          | [`script`]                       | **An array of scripts for unlocking their related input cells** (i.e. `previous_output`). See [here](https://github.com/nervosnetwork/ckb-demo-ruby-sdk/blob/develop/docs/how-to-write-contracts.md) for how to program this part. |

Reviewer 2: `unlock` is a `script`, not an array.

| `outputs`         | [`cell`]                         | **An array of cells that are used as outputs**, i.e. the newly generated cells. These are the cells may be used as inputs for other transactions. |

#### OutPoint

| Name             | Type               | Description                                                  |
| ---------------- | ------------------ | ------------------------------------------------------------ |
| `outpoint`       | {`hash` , `index`} | **An outpoint is pointer to a specific cell.** This is used in a transaction to refer a cell that is generated in a previous transaction. |
| `outpoint.hash`  | H256(hash)         | **The hash of the transaction that this cell belongs to.**   |
| `outpoint.index` | uint32             | **The index of the cell in its transaction's output list.**  |

> More information about the Transaction of Nervos CKB can be found in [whitepaper](https://github.com/nervosnetwork/rfcs/blob/afe50463bb620393b179bd8f08c263b78e366ab3/rfcs/0002-ckb/0002-ckb.md#44-transaction).

Reviewer 3: ```suggestion
More information about the Transaction of Nervos CKB can be found in [whitepaper](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0002-ckb/0002-ckb.md#44-transaction).
```

    "version": 0
},
"proposal_transactions": [],
"uncles": [
    {
    "cellbase": {
        ...
    },
    "header": {
        ...
    },
    "proposal_transactions": []
    }
]
}
```

### Description

#### Block

| Name                    | Type            | Description                                                  |
| ----------------------- | --------------- | ------------------------------------------------------------ |
| `header`                | `Header`        | **The block header of the block.** This part contains some metadata of the block. |
| `commit_trasactions`    | [`Transaction`] | **An array of transactions contained in the block.** This is where the miner put the received transactions. |
| `proposal_transactions` | [string]        | **An array of hex-encoded short transaction ID.**            |
| `uncles`                | [`UncleBlock`]  | **An array of uncle blocks of the block.**                   |

#### Header

| Name           | Type                | Description                                                  |
| -------------- | ------------------- | ------------------------------------------------------------ |

> | `cellbase_id`  | H256(hash)          | **The hash of the Cellbase transaction.** Cellbase transaction is just like the coinable transaction in Bitcoin. It's the transaction added by the miner who mined this block, by which the miner receives block reward for successfully mined the block. |

Reviewer 2: ```suggestion
| `cellbase_id`  | H256(hash)          | **The hash of the Cellbase transaction.** Cellbase transaction is just like the coinbase transaction in Bitcoin. It's the transaction added by the miner who mined this block, by which the miner receives block reward for successfully mined the block. |
```

| `difficulty`   | Bytes               | **The difficulty of the PoW puzzle.**                        |

> | `hash`         | H256(hash)          | **The block hash.**                                          |

Reviewer 2: RPC returns this, but indeed it is not a field of the header, it is computed from the header.

| `number`       | uint64              | **The block height.**                                        |
| `parent_hash`  | H256(hash)          | **The hash of the parent block.**                            |
| `seal`         | `nonce` and `proof` | **The seal of a block.** After finished the block assembling, the miner can start to do the calculation for finding the solution of the PoW puzzle. The "solution" here is called `seal`. |
| `seal.nonce`   | uint64              | **The nonce.** Similar to [the nonce in Bitcoin](https://en.bitcoin.it/wiki/Nonce). |

> | `seal.proof`   | Bytes               | **The solution of the PoW puzzle.**                          |

Reviewer 3: It's better to create a new table for `Seal` type.

Author: I don think it's necessary, same reason to the `input`.

It only increase the complexity.

| `timestamp`    | uint64              | **A [Unix time](http://en.wikipedia.org/wiki/Unix_time) timestamp.** |
| `txs_commit`   | H256(hash)          | **The Merkle Root of the Merkle trie with the hash of transactions as leaves.** |
| `txs_proposal` | H256(hash)          | **The Merkle Root of the Merkle trie with the hash of short transaction IDs as leaves.** |
| `uncles_count` | uint32              | **The number of uncle blocks.**                              |

> | `uncles_hash`  | H256(hash)          | **The Merkle Root of the Merkle trie with the hash of uncle blocks as leaves.** |

Reviewer 2: It is not the Merkle Root now.

@zhangsoledad Please provide a description for this field.

Author: The hash of serialized uncle blocks data?

Reviewer 2: This looks good to me.

Author: done

| `number`       | uint64              | **The block height.**                                        |
| `parent_hash`  | H256(hash)          | **The hash of the parent block.**                            |
| `seal`         | `nonce` and `proof` | **The seal of a block.** After finished the block assembling, the miner can start to do the calculation for finding the solution of the PoW puzzle. The "solution" here is called `seal`. |
| `seal.nonce`   | uint64              | **The nonce.** Similar to [the nonce in Bitcoin](https://en.bitcoin.it/wiki/Nonce). |
| `seal.proof`   | Bytes               | **The solution of the PoW puzzle.**                          |
| `timestamp`    | uint64              | **A [Unix time](http://en.wikipedia.org/wiki/Unix_time) timestamp.** |
| `txs_commit`   | H256(hash)          | **The Merkle Root of the Merkle trie with the hash of transactions as leaves.** |
| `txs_proposal` | H256(hash)          | **The Merkle Root of the Merkle trie with the hash of short transaction IDs as leaves.** |
| `uncles_count` | uint32              | **The number of uncle blocks.**                              |

> | `uncles_hash`  | H256(hash)          | **The hash of the serialized uncle blocks data.** This will later be changed to using Merle Trie as well. |

Reviewer 5: The serialization will be changed. Will not use Merle Trie.

Author: thx!
done
